_posts/2023-01-19-blog-new-svg-engine.md

```diff
@@ -0,0 +1,201 @@
+
+The poor performance is not the result of shortcomings in the WPE port of WebKit: when replacing the aforementioned animated SVG document fragments with HTML/CSS "equivalents" (e.g. simulating SVG circles with CSS `border-radius` tricks) the performance issue vanisheed. Why? SVG lacks support for a key feature called **accelerated compositing**, which has been available for HTML/CSS since its introduction more than a decade ago. This compositing heavily relies on the **Layer Tree**, which is unaware of SVG. Extending the **Layer Tree** implementation to account for SVG is the **main motivation for LBSE**.
+
+If you are unfamiliar with the concepts of **Render Tree** and **Layer Tree**, you might want to read the *"Key concepts"* section of an earlier <a href="https://blogs.igalia.com/nzimmermann/posts/2021-10-29-layer-based-svg-engine/page/3/#key-concepts">LBSE design document</a>, which provides an overview of the topic.
+
+### Prototyping
+
+The LBSE effort began in October 2019 as **a research project**, to find out an ideal design for the SVG **Render Tree**, that allows SVG to re-use the existing **Layer Tree** implementation (see **RenderLayer** / **RenderLayerCompositor** / **RenderLayerBacking**) with minimal changes. The aim for LBSE is to share as much code as possible with the HTML/CSS implementation, removing the need for things like SVG specific clipping/masking/filter code and disjoint HTML counterparts for the same operations.
+
+After an extensive phase of experimentation, two abandoned approaches, and a long time spent on regression fixing, the LBSE prototype was finally finished after almost two years of work. It passed all 60k+ WebKit layout tests and offered initial support for compositing, 3D transformations, `z-index`, and more. The intent was to prove that we can reach feature parity with the legacy SVG engine and retrieve the very same visual results, pixel-by-pixel (except for progressions of LBSE). Shortly after the finalization, the prototype was presented during the <a href="https://blogs.igalia.com/nzimmermann/posts/2021-10-13-svg-performance/">WebKit contributors meeting in 2021</a>.
+
+As the name "prototype" indicates, LBSE was still a prototype, impossible to upstream in the current form. It **replaces** the old SVG engine with a new one: a monolithic patch exceeding *650k* of code changes. WebKit demands small patches, with ChangeLogs, tests, etc. -- no WebKit reviewer would sign off a patch replacing a core component of WebCore. Splitting up into small pieces is also not going to work, since SVG needs to be kept intact upstream all the time. Duplicating the whole SVG engine? Not practicable either. With that problem in mind, a fruitful discussion took place with Apple during and after the WebKit contributors meeting: a realistic upstreaming strategy was defined - thanks *Simon Fraser* for suggesting a pragmatic approach!
```

Comment:
Makes sense. OMG 650k code changes would be too much, its kb :-)

I've adapted the text:

```diff
-As the name "prototype" indicates, LBSE was still a prototype, impossible to upstream in the current form. It **replaces** the old SVG engine with a new one: a monolithic patch exceeding *650k* of code changes. WebKit demands small patches, with ChangeLogs, tests, etc. -- no WebKit reviewer would sign off a patch replacing a core component of WebCore. Splitting up into small pieces is also not going to work, since SVG needs to be kept intact upstream all the time. Duplicating the whole SVG engine? Not practicable either. With that problem in mind, a fruitful discussion took place with Apple during and after the WebKit contributors meeting: a realistic upstreaming strategy was defined - thanks *Simon Fraser* for suggesting a pragmatic approach!
+As the name "prototype" indicates, LBSE was not ready for integration into WebKit at this point. It **replaced** the old SVG engine with a new one, resulting in a monolithic patch exceeding *650 kb* of code changes. WebKit demands small patches, with ChangeLogs, tests, etc. -- no WebKit reviewer would sign off a patch replacing a core component of WebCore in one shot. Splitting up into small pieces is also not going to work, since SVG needs to be kept intact upstream all the time. Duplicating the whole SVG engine? Not practicable either. With that problem in mind, a fruitful discussion took place with Apple during and after the WebKit contributors meeting: a realistic upstreaming strategy was defined - thanks *Simon Fraser* for suggesting a pragmatic approach!
```

Reply:
Pushed this fix - please mark as resolved if it's fine.